From 0db9f49bc9c8018eb02103c203b1d4aefbac6c66 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 15 Mar 2018 10:32:34 +0900
Subject: [PATCH 307/909] drm: rcar-du: Fix DPLL workaround for H3 ES1.x

commit b49fc1b1ff84aa287df73df9e32f7b8027b5753f from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

DPLL does not function unless 20 bit of DPLLCR register is set
in H3 ES1.x. According to latest H/W manual, 20 bit is reserved bit.
This patch uses soc_device_match to separate bit settings
for each H/W revision.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c | 10 ++++++++--
 drivers/gpu/drm/rcar-du/rcar_du_regs.h |  3 ++-
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index a39167c8b735..55d00d646d9b 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -277,13 +277,19 @@ static void rcar_du_crtc_set_display_timing(struct rcar_du_crtc *rcrtc)
 					   | DPLLCR_N(dpll.n) | DPLLCR_M(dpll.m)
 					   | DPLLCR_STBY;
 
-				if (rcrtc->index == 1)
+				if (rcrtc->index == 1) {
 					dpllcr |= DPLLCR_PLCS1
 					       |  DPLLCR_INCS_DOTCLKIN1;
-				else
+				} else {
 					dpllcr |= DPLLCR_PLCS0
 					       |  DPLLCR_INCS_DOTCLKIN0;
 
+					if (soc_device_match(
+							rcar_du_r8a7795_es1))
+						dpllcr |=
+						    DPLLCR_PLCS0_H3ES1X_WA;
+				}
+
 				rcar_du_group_write(rcrtc->group, DPLLCR,
 						    dpllcr);
 			}
diff --git a/drivers/gpu/drm/rcar-du/rcar_du_regs.h b/drivers/gpu/drm/rcar-du/rcar_du_regs.h
index 9dfd220ceda1..1b1390e456ef 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_regs.h
+++ b/drivers/gpu/drm/rcar-du/rcar_du_regs.h
@@ -285,7 +285,8 @@
  * isn't implemented by other SoC in the Gen3 family it can safely be set
  * unconditionally.
  */
-#define DPLLCR_PLCS0		(3 << 20)
+#define DPLLCR_PLCS0		(1 << 21)
+#define DPLLCR_PLCS0_H3ES1X_WA	(1 << 20)
 #define DPLLCR_CLKE		(1 << 18)
 #define DPLLCR_FDPLL(n)		((n) << 12)
 #define DPLLCR_N(n)		((n) << 5)
-- 
2.17.1

